### ISSUE TITLE
SMTP login not working for hMailServer

### ISSUE THREAD

**Issue by mbaeten** [deltachat/deltachat-android#67]
Hi there, just tested delta chat and it seems to have some hickups with hmailserver (SMTP at least). IMAP seems to work fine.

* Version: 0.1.25 (f-droid)
* Server: hMailServer
* Tested with ssl, starttls and without encryption

Delta chat error message:
Error: SMTP-login failed for user **** (Login failed)

Server logs:
"SMTPD"	2808	****	"2017-04-10 20:56:55.044"	"**.**.**.**"	"SENT: 220 mail.*********.***"
"SMTPD"	2772	****	"2017-04-10 20:56:55.076"	"**.**.**.**"	"RECEIVED: EHLO localhost"
"SMTPD"	2772	****	"2017-04-10 20:56:55.076"	"**.**.**.**"	"SENT: 250-mail.*********.***[nl]250-SIZE 258260000[nl]250-STARTTLS[nl]250-AUTH LOGIN[nl]250 HELP"
"SMTPD"	2792	****	"2017-04-10 20:56:55.123"	"**.**.**.**"	"RECEIVED: STARTTLS"
"SMTPD"	2792	****	"2017-04-10 20:56:55.138"	"**.**.**.**"	"SENT: 220 Ready to start TLS"
"SMTPD"	2772	****	"2017-04-10 20:56:55.341"	"**.**.**.**"	"RECEIVED: EHLO localhost"
"SMTPD"	2772	****	"2017-04-10 20:56:55.341"	"**.**.**.**"	"SENT: 250-mail.*********.***[nl]250-SIZE 258260000[nl]250-AUTH LOGIN[nl]250 HELP"
"SMTPD"	2808	****	"2017-04-10 20:56:55.388"	"**.**.**.**"	"RECEIVED: QUIT"
"SMTPD"	2808	****	"2017-04-10 20:56:55.388"	"**.**.**.**"	"SENT: 221 goodbye"

Regards :-)

**Comment by r10s** [comment]
Can you give me an test account? This would make testing much easier.

**Comment by mbaeten** [comment]
Done

**Comment by r10s** [comment]
Thank you very much, I will check this out as soon as possible.

**Comment by r10s** [comment]
I've just tested the account with SSL and first, sending did not work. Delta Chat did not got an error message but the mails were not send.

Some minutes later, mailing works just fine. Hmmm.

Maybe you can see some stuff in the error log?

I'll try this again these days, maybe we get the point.

(NB: Receiving works all the time, STARTTLS not tested)

**Comment by schmidtea** [comment]
I'm having the exact same issue - running hmailserver, Delta Chat saying it can't login to the smtp server. Tried both with and without SSL/TLS encryption; same result.

It seems that the error is in the SMTP handshake. DeltaChat never tries to login; it receives the response to its EHLO, and then proceeds to send QUIT.

**Comment by r10s** [comment]
Okay, so, Delta Chat does not login to SMTP reproducible? This sounds debuggable. Can you give me a test account? ( r10s@b44t.com )

**Comment by schmidtea** [comment]
I've set up a test account for you. Details in your e-mail.

**Comment by r10s** [comment]
Thank you very much, I will test it these days.

**Comment by r10s** [comment]
Hmm, I tested the account using the linux CLI program - no problem.
I will try over with the Android App ...

**Comment by r10s** [comment]
Okay, I can reproduce the error with the Android App ... seems as if the reason for the difference to the CLI program are different libEtPan versions. 

I will check this these days - can you please leave the testing account? Thank you very much!

**Comment by schmidtea** [comment]
OK - good that it's reproducible. I'll leave the testing account up for now. Please keep me posted.

**Comment by Yetangitu** [comment]
I can add to this that SMTP login also gives problems on Exim, at least for me. Exim (4.88-2 Debian Sid) tells me that...

```
cram_md5_server authenticator failed for mail.example.org (localhost) [192.168.1.17]: 535 Incorrect authentication data (set_id=username)` while Delta says `Error: SMTP-login failed for user username (Login failed)
```

In the Android (CM11, Android 4.4.4) log Delta seems to think it succeeded:

```
/DeltaChat(12803): T3: Connecting to SMTP-server "mail.example.org:587"...
I/DeltaChat(12803): T4: Attaching ok.
```

...which is, of course, untrue.

Using STARTTLS on port 587.

**Comment by r10s** [comment]
Anyone who can fix this? I fear, I have no time for this currently.

**Comment by johanw666** [comment]
This issue is open for over a year now. I recently came into contact with DeltaChat but I have the same issue which makes DC useless for me, is there any progress on this? In the anonymized logfile from the server I find this on the failed SMTP attempt:
```
"DEBUG"	2724	"2018-05-03 20:33:25.301"	"Creating session 186960"
"TCPIP"	2724	"2018-05-03 20:33:25.301"	"TCP - w.x.y.z connected to a.b.c.d:25."
"DEBUG"	2724	"2018-05-03 20:33:25.301"	"TCP connection started for session 186959"
"SMTPD"	2724	186959	"2018-05-03 20:33:25.301"	"113.105.200.40"	"SENT: 220 mydomain.com ESMTP"
"SMTPD"	2752	186959	"2018-05-03 20:33:25.576"	"113.105.200.40"	"RECEIVED: EHLO sjsedau.com"
"SMTPD"	2752	186959	"2018-05-03 20:33:25.576"	"113.105.200.40"	"SENT: 250-mydomain.nl[nl]250-SIZE[nl]250-AUTH LOGIN PLAIN[nl]250 HELP"
"SMTPD"	2724	186959	"2018-05-03 20:33:26.431"	"113.105.200.40"	"RECEIVED: AUTH LOGIN"
"SMTPD"	2724	186959	"2018-05-03 20:33:26.431"	"113.105.200.40"	"SENT: 334 VXNlcm5hbWU6"
"SMTPD"	2740	186959	"2018-05-03 20:33:26.666"	"113.105.200.40"	"RECEIVED: NEUyNDlGRDQuOTA3MDYwM0B2dWxjYW4ueHM0YWxsLm5s"
"SMTPD"	2740	186959	"2018-05-03 20:33:26.666"	"113.105.200.40"	"SENT: 334 UGFzc3dvcmQ6"
"SMTPD"	2724	186959	"2018-05-03 20:33:26.941"	"113.105.200.40"	"RECEIVED: ***"
"SMTPD"	2724	186959	"2018-05-03 20:33:26.946"	"113.105.200.40"	"SENT: 535 Authentication failed. Restarting authentication process."
"DEBUG"	2748	"2018-05-03 20:33:27.181"	"The read operation failed. Bytes transferred: 0 Remote IP: w.x.y.z, Session: 186959, Code: 2, Message: End of file"
"DEBUG"	2748	"2018-05-03 20:33:27.181"	"Ending session 186959"
```

**Comment by johanw666** [comment]
Any idea when 0.17 is released? I tried building it myself (downloaded the zipfile and opened it in Android Studio) but building a release apk without downloading the core libraries succeeds and results in a small apk of only 3.5MB so I think I'm doing something wrong. I'll study the instructions on the front page.

**Comment by johanw666** [comment]
0.17 did not solve the issue, I see exactly the same now in my logfiles:
```
"DEBUG"	2760	"2018-05-08 20:06:09.761"	"TCP connection started for session 279829"
"DEBUG"	2760	"2018-05-08 20:06:09.761"	"Performing SSL/TLS handshake for session 279829. Verify certificate: False"
"TCPIP"	2720	"2018-05-08 20:06:09.839"	"TCPConnection - TLS/SSL handshake completed. Session Id: 279829, Remote IP: 82.161.223.137, Version: TLSv1.2, Cipher: ECDHE-RSA-AES256-GCM-SHA384, Bits: 256"
"SMTPD"	2720	279829	"2018-05-08 20:06:09.839"	"82.161.223.137"	"SENT: 220 vulcan.xs4all.nl ESMTP"
"SMTPD"	2764	279829	"2018-05-08 20:06:09.901"	"82.161.223.137"	"RECEIVED: EHLO localhost"
"SMTPD"	2764	279829	"2018-05-08 20:06:09.901"	"82.161.223.137"	"SENT: 250-vulcan.xs4all.nl[nl]250-SIZE[nl]250-AUTH LOGIN PLAIN[nl]250 HELP"
"SMTPD"	2720	279829	"2018-05-08 20:06:09.979"	"82.161.223.137"	"RECEIVED: AUTH PLAIN ZGMxAGRjMQBgQXVyb3JhJw=="
"SMTPD"	2720	279829	"2018-05-08 20:06:09.979"	"82.161.223.137"	"SENT: 535 Authentication failed. Restarting authentication process."
"SMTPD"	2732	279829	"2018-05-08 20:06:09.995"	"82.161.223.137"	"RECEIVED: QUIT"
"SMTPD"	2732	279829	"2018-05-08 20:06:09.995"	"82.161.223.137"	"SENT: 221 goodbye"
"DEBUG"	2720	"2018-05-08 20:06:09.995"	"Ending session 279829"
```

**Comment by r10s** [comment]
@johanw666 thank you for the information, now we know this.
do you have the possibilities to debug a running Delta Chat while watching at the same time what is going on on the hMailServer? Or anyone else?
can anyone confirm that error does _not_ pop up in the Delta/CLI? If so, there may be more differences between the OpenSSL/Android OpenSSL/System/CLI than @Ampli-fier found out.

**Comment by Ampli-fier** [comment]
I can do the debugging in Delta Chat, both Android and Core. I think I don't even need the server logs.

@johanw666 Can you provide a test account to @r10s ? And provide the access details (Protocols, Ports)?

**Comment by johanw666** [comment]
@r10s since I can't yet get it to build completely I can't run a debug console in Android Studio. I can get logcat output it that is sufficient. I mailed the data for a test account to @r10s (plaintext because I could not find a pgp key from him).

**Comment by r10s** [comment]
@johanw666 thank you very much, i forwarded the account data to @Ampli-fier (i think this was the idea? not sure :)

**Comment by r10s** [comment]
> plaintext because I could not find a pgp key

my hope is that some time in the future this is no longer in issue because e-mail clients support Autocrypt by default and users just have the keys without even thinking about it </dream mode>

**Comment by Ampli-fier** [comment]
Got the account details, thanks. Configured the account in K9 and will start with Delta Chat soon.

**Comment by r10s** [comment]
@Ampli-fier great that you push this forward :)

**Comment by Ampli-fier** [comment]
@johanw666 
You can see some failed and some successful SMTP logins to your server.

Currently it looks to me that there is a strange behaviour in Cyrussasl, in _sasl_encode64_.
This gets called here: [mailsmtp.c](https://github.com/dinhviethoa/libetpan/blob/d4244c326e392e7c581383d088e302b4fcb3d921/src/low-level/smtp/mailsmtp.c#L1389)
and should return `"\0username\0password"` base64-encoded, but I see `"\0username\0username\0password"` instead.

When I _inject_ the correct syntax (remove the 2nd username occurrence) in `mailsmtp.c` then I can login with Delta Chat Core (did not try with Android).

I need to check the coming days where this behaviour comes from / if I did all correctly ...

**Comment by johanw666** [comment]
@r10s I mailed you that in a separate mail when I saw I didn't tell it - it's just username@server

**Comment by Ampli-fier** [comment]
> happy vatertag :)

Thanks, same to you - it was a busy day :)

**Comment by Ampli-fier** [comment]
When I omit the _logical and_ part: `user && *user`
in  cyrus-sasl/plugins/plain.c lines [400](https://github.com/cyrusimap/cyrus-sasl/blob/e1ddba9094de241bfcb50c646cd505e34cd14cf5/plugins/plain.c#L400) and [411](https://github.com/cyrusimap/cyrus-sasl/blob/e1ddba9094de241bfcb50c646cd505e34cd14cf5/plugins/plain.c#L411) then SMTP login works with Android.

@r10s Can you please have a quick look at this file? What is the reason this test results in 'true'?

**Comment by Ampli-fier** [comment]
I think the user entry became optional with [this](https://github.com/cyrusimap/cyrus-sasl/commit/4454402df874503b3e1327db6534ae0aca8cc4c0) commit. Gonna try this later ... tonight.

**Comment by r10s** [comment]
`user && *user` is true if `user` is not a NULL-pointer and not of length 0. 

- the first part checks if the pointer _is not NULL_. 
- if it _is not NULL_, the second part dereferences the pointer and checks that the referenced character is not 0 - if the references character is null, this is the terminating null-character of the string and therefore the string is of length 0.

@Ampli-fier Please ping me, if this is not understandable or you need more info

**Comment by Ampli-fier** [comment]
I tend to think SASL is okay, Libetpan is simply providing the `user` argument twice to Cyrus-SASL.

This can be seen [here](https://github.com/dinhviethoa/libetpan/blob/d2b340bd57229976f00cecfdef909cd77b90303e/src/low-level/smtp/mailsmtp.c#L1032).

Modifying this part from
```
return mailesmtp_auth_sasl(session, "PLAIN", hostname, NULL, NULL, user, user, pass, NULL);
```
to
```
return mailesmtp_auth_sasl(session, "PLAIN", hostname, NULL, NULL, NULL, user, pass, NULL);
```
results in successful SMTP login procedure.

K-9 does it the same way (only user + password), see [SmtpTransport.java
](https://github.com/k9mail/k-9/blob/e80aa401defbe7b0eab0be7012b0cd8b4971feba/k9mail-library/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.java#L736)

**Comment by Ampli-fier** [comment]
[This discussion](https://www.hmailserver.com/forum/viewtopic.php?f=7&t=23500&p=141945&hilit=PLAIN+SASL#p141945) on the hmailserver forum is about _our_ topic.

One user stated:
```
Looks like server does not support authorization identities and implements
only basic 'authentication identity + password' AUTH command.

If it causes problems, disable plain text authentication in hMailServer settings
and let your programs use AUTH LOGIN instead.

Register bug report if you like. hMailServer 5.4 B1944 still has same problem.
```
@johanw666 
Can you try to disable PLAIN authentication in your server? I made a check with a patched version of Delta Chat which used LOGIN even though libetpan decided to use PLAIN ... and succeeded.

BTW. Which client(s) do you use to send mails with your server? I checked some repos here on Github and all of them using libetpan do it "the Delta Chat way" ;-)

**Comment by johanw666** [comment]
I use my server with Thunderbird, K9 on Android and a family member uses it with the standard mail clientfrom her iPhone. They all work. In K9 I use SSL/TLS, not STARTTLS, which is also in Delta Chat an option.

When I disable "Allow plaintext authentication" in hMailServer (settings - protocols - SMTP - RFC compliance tab) Delta Chat registers OK. K9 and Thunderbird also still work. I'll have to ask the iPhone user this weekend if that also still works. For now I'll leave plaintext login disabled.

**Comment by johanw666** [comment]
OK, I can login now, send messages and I see them comming inon the account on Thunderbird. But the other phone does not receive them. I'll have to check that tomorrow.

**Comment by Ampli-fier** [comment]
I was looking for the reasons why the clients you use worked fine with PLAIN authentication method:

1.  K-9 sends only user + password, see [SmtpTransport.java
](https://github.com/k9mail/k-9/blob/e80aa401defbe7b0eab0be7012b0cd8b4971feba/k9mail-library/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.java#L736)
2. Thunderbird sends only user + password (excerpt from `nsSmtpProtcol.cpp`):
```
  else if (m_currentAuthMethod == SMTP_AUTH_PLAIN_ENABLED)
  {
    MOZ_LOG(SMTPLogModule, mozilla::LogLevel::Debug, ("PLAIN auth"));
    char plain_string[512];
    int len = 1; /* first <NUL> char */

    memset(plain_string, 0, 512);
    PR_snprintf(&plain_string[1], 510, "%s", username.get());
    len += username.Length();
    len++; /* second <NUL> char */
    PR_snprintf(&plain_string[len], 511-len, "%s", uniPassword.get());
    len += password.Length();

    base64Str = PL_Base64Encode(plain_string, len, nullptr);
    PR_snprintf(buffer, sizeof(buffer), "AUTH PLAIN %.256s" CRLF, base64Str);
  }
```
3. Apple: In the hmailserver forum entry mentioned above someone stated:
```
... the mac client logs in according to the rfc using the format

authzid UTF8NUL authcid UTF8NUL passwd

If this is rejected (and hMailServer rejects it)

the mac tries again with

UTF8NUL authcid UTF8NUL passwd 
```

**Comment by r10s** [comment]
okay - but it's still a problem with Delta Chat, isn't it?
If i understand correctly, Delta Chat does not work with the defaults of hMailServer while other (K-9, Thunderbird) do?

@johanw666 Wondering if you have set "Allow plaintext authentication" to the original value that makes problems with Delta (maybe i find finally some time to check this out myself ... but it would not make much sense when the server-workaround is enabled)

Also wondering how complicated it is to setup an own hMailServer - does it work in VirtualBox? (i do not really have a Windows system here)

**Comment by johanw666** [comment]
@r10s No I have removed the option to make Delta Chat work. I can re-enable it if you inform me you want to work on it. Wether it runs in VirtualBox, I don't know.  I don't see why not, probably the port forwardings reuired to mke it work are the most complicated to set up. It's quite easy to setup - just one executable, run it and create a mail domain and accounts in the manegement console. After comming from `sendmail.conf` on Linux it was dead easy. :)

**Comment by t-oster** [comment]
I just installed a fresh hMailServer in VirtualBox (yes it works in VirtualBox btw). It worked out of the box with the current Android version (after I manually set the IP of the server, because I don't have DNS set up and after I set security to Off since I did not set up an SSL certificate).

After I enabled Plaintext Auth (which is DISABLED by default) an restarted the Server, it stopped working.

So if I understand correctly: I we update libetpan to only send username+password, it will work. And since other mail clients do it the same way, it should no break anything? Otherwise we could update libetpan to try again with only username+password if the previous attempt fails?

Or we even use the `mailesmtp_auth_sasl` function directly from DC in that case so we do not need to patch libetpan.
